http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java

```diff
@@ -0,0 +1,180 @@
+            long responseSize,
+            int responseCode,
+            long timeToLastByte,
+            Optional<String> failureReason)
+    {
+        this.timeStamp = requireNonNull(timeStamp, "timeStamp is null");
+        this.protocolVersion = requireNonNull(protocolVersion, "protocolVersion is null");
+        this.method = requireNonNull(method, "method is null");
+        this.requestUri = requireNonNull(requestUri, "requestUri is null");
+        this.traceToken = traceToken;
+        this.responseSize = responseSize;
+        this.responseCode = responseCode;
+        this.timeToLastByte = timeToLastByte;
+        this.failureReason = requireNonNull(failureReason, "failureReason is null");
+    }
+
+    public Instant getTimeStamp()
+    {
+        return timeStamp;
+    }
+
+    public String getProtocolVersion()
+    {
+        return protocolVersion;
+    }
+
+    public String getMethod()
+    {
+        return method;
+    }
+
+    public String getRequestUri()
+    {
+        return requestUri;
+    }
+
+    public String getTraceToken()
+    {
+        return traceToken;
+    }
+
+    public long getResponseSize()
+    {
+        return responseSize;
+    }
+
+    public String getResponseCode()
+    {
+        return failureReason.orElseGet(() -> Integer.toString(responseCode));
+    }
+
+    public long getTimeToLastByte()
+    {
+        return timeToLastByte;
+    }
+
+    static HttpRequestEvent createHttpRequestEvent(RequestInfo requestInfo, ResponseInfo responseInfo)
+    {
+        requireNonNull(requestInfo, "requestInfo is null");
+        requireNonNull(responseInfo, "responseInfo is null");
+        Request request = requestInfo.getRequest();
+        Optional<Response> response = responseInfo.getResponse();
+
+        String requestUri = null;
+        if (request.getURI() != null) {
+            requestUri = request.getURI().toString();
+        }
+
+        String method = request.getMethod();
+        if (method != null) {
+            method = method.toUpperCase(Locale.US);
+        }
+
+        long responseSize = NO_RESPONSE;
+        int responseCode = NO_RESPONSE;
+        if (response.isPresent()) {
+            responseSize = responseInfo.getResponseSize();
+            responseCode = response.get().getStatus();
+        }
+
+        long timeToLastByte = max(responseInfo.getResponseTimestampMillis() - requestInfo.getRequestTimestampMillis(), 0L);
+
+        return new HttpRequestEvent(
+                Instant.ofEpochMilli(requestInfo.getRequestTimestampMillis()),
+                request.getVersion().toString(),
+                method,
+                requestUri,
+                getHeader(request, TRACETOKEN_HEADER),
+                responseSize,
+                responseCode,
+                timeToLastByte,
+                getFailureReason(responseInfo));
+    }
+
+    @VisibleForTesting
+    static Optional<String> getFailureReason(ResponseInfo responseInfo)
```

Comment:
What are possible failure reasons? Have you looked at whether the common cases produces something reasonable given this logic?

Reply:
I remember seeing TIMEOUT, CANCELLATION, and IO in the logs.

Reply:
So, these are probably good for debugging many cases. If we find a better way we can improve it in the future.